File: rutt.py
#!/usr/bin/env python

import argparse
import calendar
import curses
import curses.wrapper
import datetime
import feedparser
import os
import sqlalchemy
import sqlite3
import string
import sys
import time
import webbrowser

from elixir import *

def printable(input):
    """
    Removes unicode characters so they can be displayed on Terminal.
    """
    return ''.join([x for x in input if x in string.printable])

class Feed(Entity):
    id = Field(Integer, primary_key=True)
    items = OneToMany('FeedItem')

    title = Field(String(255))
    url = Field(String(255), unique=True)
    update_interval = Field(Integer, default=3600)

    created_at = Field(DateTime, default=datetime.datetime.now)
    updated_at = Field(DateTime, default=datetime.datetime.now)


    def __init__(self, url):
        """
        Given a url add this feed to the database.
        """

        rss = feedparser.parse(url)

        self.title = rss.feed.title
        self.url = url

    def refresh(self):
        rss = feedparser.parse(self.url)

        for item in rss.entries:
            try:
                feed_item = FeedItem(self, item)
                session.commit()
            except:
                session.rollback()

    def unread(self):
        return len([1 for item in self.items if item.is_read == False])

    def remove(self):
        try:
            for item in self.items:
                item.delete()
            self.delete()
            session.commit()
        except:
            session.rollback()

class FeedItem(Entity):
    id = Field(Integer, primary_key=True)
    feed = ManyToOne('Feed')

    title = Field(String(255))
    url = Field(String(255), unique=True)

    description = Field(Text)

    is_read = Field(Boolean, default=False)
    like_this = Field(Boolean, default=False)

    published_at = Field(DateTime, default=datetime.datetime.now)

    created_at = Field(DateTime, default=datetime.datetime.now)
    updated_at = Field(DateTime, default=datetime.datetime.now)

    using_options(order_by='-published_at')

    def __init__(self, feed, item):
        """
        Create a new item for a feed.
        """
        self.feed = feed

        self.title = item.title
        self.url = item.link

        if item.has_key('published'):
            self.published_at = datetime.datetime(*item.published_parsed[:6])
        elif item.has_key('updated'):
            self.published_at = datetime.datetime(*item.updated_parsed[:6])
        elif item.has_key('created'):
            self.published_at = datetime.datetime(*item.published_parsed[:6])

    def mark_as_read(self):
        """
        Mark the current item as read.
        """
        self.is_read = True
        session.commit()

    def mark_as_unread(self):
        """
        Mark the current item as unread.
        """
        self.is_read = False
        session.commit()


class Screen(object):
    def __init__(self, stdscr):
        self.stdscr = stdscr

        self.min_y = 1
        self.max_y = curses.LINES

        self.cur_y = 1
        self.cur_x = 0

    def display_menu(self):
        self.stdscr.clear()
        self.stdscr.addstr(0, 0, " rutt %s\n" % self.menu, curses.A_REVERSE)

    def move_pointer(self, pos, move_to=False):
        self.stdscr.addstr(self.cur_y, 0, " ")

        if move_to is True:
            self.cur_y = pos
        else:
            self.cur_y += pos

        self.stdscr.addstr(self.cur_y, 0, ">", curses.A_REVERSE)


class FeedScreen(Screen):
    def __init__(self, stdscr):
        self.feeds = {}
        self.menu = "q:Quit d:delete"

        self.reload_feeds()

        super(FeedScreen, self).__init__(stdscr)

    def reload_feeds(self):
        self._feeds = Feed.query.all()

    def display_feeds(self):
        self.cur_y = self.min_y

        for feed in self._feeds[self.limit[0]:self.limit[1]]:
            if feed.unread() == 0:
                continue

            self.stdscr.addstr(self.cur_y, 0, "  %d/%d\t\t%s\n" % (feed.unread(), len(feed.items), printable(feed.title),))
            self.feeds[self.cur_y] = feed

            self.cur_y += 1

        self.cur_y = self.min_y
        self.stdscr.refresh()

    def window(self, start_limit=None, end_limit=None):
        self.stdscr.clear()

        if start_limit or end_limit:
            self.limit = (start_limit, end_limit)

        self.display_menu()
        self.display_feeds()
        self.move_pointer(0)

    def loop(self):
        self.window(0, curses.LINES - 2)

        while True:
            c = self.stdscr.getch()
            if 0 < c < 256:
                if chr(c) in 'Qq':
                    break
                elif chr(c) in 'Aa':
                    pass
                elif chr(c) in 'Dd':
                    cur_y = self.cur_y
                    self.stdscr.clear()
                    self.display_menu()
                    feed = self.feeds[cur_y]
                    self.stdscr.addstr(2, 0, "Are you sure you want to delete %s? " % printable(feed.title))
                    d = self.stdscr.getch()
                    if chr(d) in 'Yy':
                        feed.remove()
                        self.reload_feeds()
                    self.window()
                    self.move_pointer(cur_y, move_to=True)
                elif chr(c) in 'Pp':
                    self.window(self.limit[0] - curses.LINES - 2, self.limit[0])
                elif chr(c) in 'Nn':
                    self.window(self.limit[1], self.limit[1] + curses.LINES - 2)
                elif chr(c) == ' ':
                    cur_y = self.cur_y
                    item_screen = ItemScreen(self.stdscr, self.feeds[cur_y])
                    item_screen.loop()

                    self.window()
                    self.move_pointer(cur_y, move_to=True)
            else:
                if c == curses.KEY_UP:
                    self.move_pointer(-1)
                elif c == curses.KEY_DOWN:
                    self.move_pointer(1)

class ItemScreen(Screen):
    def __init__(self, stdscr, feed):
        self.feed = feed
        self.items = {}
        self.menu = " i:quit r:refresh m:mark as read u:mark as unread"

        super(ItemScreen, self).__init__(stdscr)

    def display_items(self):
        self.cur_y = self.min_y

        for item in self.feed.items[self.limit[0]:self.limit[1]]:
            self.stdscr.addstr(self.cur_y, 0, "  %s\t%s\t%s\n" % ('N' if not item.is_read else ' ', item.published_at, printable(item.title),))

            self.items[self.cur_y] = item
            self.cur_y += 1

        self.cur_y = self.min_y
        self.stdscr.refresh()

    def window(self, start_limit=None, end_limit=None):
        self.stdscr.clear()

        if start_limit or end_limit:
            self.limit = (start_limit, end_limit)

        self.display_menu()
        self.display_items()
        self.move_pointer(0)

    def loop(self):
        self.window(0, curses.LINES - 2)

        while True:
            c = self.stdscr.getch()
            if 0 < c < 256:
                if chr(c) in 'IiQq':
                    break
                elif chr(c) in 'Pp':
                    self.window(self.limit[0] - curses.LINES - 2, self.limit[0])
                elif chr(c) in 'Nn':
                    self.window(self.limit[1], self.limit[1] + curses.LINES - 2)
                elif chr(c) in 'Mm':
                    cur_y = self.cur_y
                    self.items[cur_y].mark_as_read()
                    self.window()
                    self.move_pointer(cur_y, move_to=True)
                elif chr(c) in 'Uu':
                    cur_y = self.cur_y
                    self.items[cur_y].mark_as_unread()
                    self.window()
                    self.move_pointer(cur_y, move_to=True)
                elif chr(c) in 'Rr':
                    self.feed.refresh()
                    self.window()
                elif chr(c) == ' ':
                    content_screen = ContentScreen(self.stdscr, self.items[self.cur_y])
                    content_screen.loop()

                    self.window()
            else:
                if c == curses.KEY_UP:
                    self.move_pointer(-1)
                elif c == curses.KEY_DOWN:
                    self.move_pointer(1)



class ContentScreen(Screen):
    def __init__(self, stdscr, item):
        self.item = item
        self.menu = "i:back b:open in browser"

        super(ContentScreen, self).__init__(stdscr)

    def get_content(self):
        render_cmd = "elinks -dump -dump-charset ascii -force-html %s" % self.item.url
        self.content = printable(os.popen(render_cmd).read()).split("\n")

    def move_pointer(self, pos):
        if self.cur_line + pos < 0:
            return

        self.stdscr.addstr(1, 2, "%s (%s)\n" % (self.item.title, self.item.url), curses.A_BOLD)

        self.cur_line = self.cur_line + pos

        lines = self.content[self.cur_line:self.cur_line + curses.LINES - 5]
        cur_y = 2
        for line in lines:
            self.stdscr.addstr(cur_y, 2, "%s\n" % line)
            cur_y += 1

        self.stdscr.refresh()

    def window(self, pointer_pos):
        self.stdscr.clear()
        self.display_menu()
        self.move_pointer(pointer_pos)

    def loop(self):
        self.cur_line = 0
        self.get_content()
        self.window(0)

        while True:
            c = self.stdscr.getch()
            if 0 < c < 256:
                if chr(c) in 'IiQq':
                    self.item.mark_as_read()
                    break
                elif chr(c) in 'Bb':
                    webbrowser.open_new_tab(self.item.url)
                elif chr(c) in ' ':
                    self.window(10)
            else:
                if c == curses.KEY_UP:
                    self.window(-1)
                elif c == curses.KEY_DOWN:
                    self.window(1)

def start_screen():
    stdscr = curses.initscr()

    curses.start_color()
    curses.init_pair(1, curses.COLOR_GREEN, curses.COLOR_BLACK)
    curses.init_pair(2, curses.COLOR_YELLOW, curses.COLOR_BLACK)
    curses.init_pair(3, curses.COLOR_RED, curses.COLOR_BLACK)

    curses.noecho()
    curses.cbreak()
    stdscr.keypad(1)

    stdscr.clear()

    feed_screen = FeedScreen(stdscr)
    feed_screen.loop()

    # Die!
    stdscr.keypad(0)
    curses.echo()
    curses.nocbreak()
    curses.endwin()

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="rutt - Mutt like RSS/Atom reader")
    parser.add_argument('-a', '--add', nargs='+', help="Add a new feed.", metavar="url")
    parser.add_argument('-r', '--reload', action='store_true', help="Update feeds.")
    parser.add_argument('-l', '--list-feeds', action='store_true', help="List the feeds")
    args = parser.parse_args()

    metadata.bind = "sqlite:///rutt.db"
    metadata.bind.echo = False

    # Create the models.
    setup_all()
    create_all()

    if args.add is not None:
        for url in args.add:
            try:
                feed = Feed(url)
                session.commit()
            except sqlalchemy.exc.IntegrityError:
                session.rollback()
            except:
                print "Failed to add %s" % url

        sys.exit()

    if args.list_feeds is True:
        for feed in Feed.query.all():
            print feed.url
        sys.exit()

    if args.reload is True:
        for feed in Feed.query.all():
            feed.refresh()

        sys.exit()

    start_screen()

File: rutt.rb
require 'rubygems'

require 'launchy'
require 'ncurses'
require 'nokogiri'
require 'open-uri'
require 'optparse'
require 'rss/1.0'
require 'rss/2.0'
require 'sqlite3'
require 'feedparser'

# gem install launchy
# gem install sqlite3
# gem install ncurses
# gem install ruby-feedparser

$db = SQLite3::Database.new('rutt.db')
$db.results_as_hash = true

module Config

end

module Opml
  def self.get_urls(file)
    doc = Nokogiri::XML(open(file))

    urls = []

    doc.xpath('opml/body/outline').each do |outline|
      if outline['xmlUrl']
        urls << outline['xmlUrl']
      else
        (outline/'outline').each do |outline2|
          urls << outline2['xmlUrl']
        end
      end
    end

    return urls
  end
end

module Feed
  extend self

  def make_table!
    $db.execute(%{
         create table if not exists feeds (
                      id integer PRIMARY KEY,
                      title text,
                      url text,
                      update_interval integer default 3600,
                      created_at NOT NULL DEFAULT CURRENT_TIMESTAMP,
                      updated_at NOT NULL DEFAULT CURRENT_TIMESTAMP,
                      UNIQUE(url))
      })
  end

  def add(url, refresh=true)
    $db.execute("insert or ignore into feeds (url) values ('#{url}')")
    $db.execute("select * from feeds where id = (select last_insert_rowid())") do |feed|
      refresh_for(feed)
    end if refresh
  end

  def get(id)

  end

  def delete(feed)
    $db.execute("delete from items where feed_id = ?", feed['id'])
    $db.execute("delete from feeds where id = ?", feed['id'])
  end

  def all
    $db.execute(%{
       select f.*,
              (select count(*) from items iu where iu.feed_id = f.id) as num_items,
              (select count(*) from items ir where ir.read = 0 and ir.feed_id = f.id) as unread
       from feeds f
    })
  end

  def refresh
    $db.execute("select * from feeds") do |feed|
      refresh_for(feed)
    end
  end

  def refresh_for(feed)
    content = open(feed['url']).read
    rss = FeedParser::Feed::new(content)

    $db.execute("update feeds set title = ? where id = ?", rss.title, feed['id'])

    rss.items.each do |item|
      $db.execute("insert or ignore into items (feed_id, title, url, published_at) values (?, ?, ?, ?)", feed['id'], item.title, item.link, item.date.to_i)
    end
  end

  def unread(feed_id)
    $db.execute("select * from items where feed_id = ? and read = 0", feed)
  end

  def mark_as_read(feed)
    $db.execute("update items set read = 1 where feed_id = ?", feed['id'])
  end
end


module Item

  extend self

  def make_table!
    $db.execute(%{
      create table if not exists items (
                      id integer PRIMARY KEY,
                      feed_id integer,
                      title text,
                      url text,
                      description text,
                      read int default 0,
                      prioritize int default 0,
                      published_at DATE NOT NULL DEFAULT (datetime('now','localtime')),
                      created_at NOT NULL DEFAULT CURRENT_TIMESTAMP,
                      updated_at NOT NULL DEFAULT CURRENT_TIMESTAMP,
                      UNIQUE(url),
                      FOREIGN KEY(feed_id) REFERENCES feeds(id))
    })
  end

  def all(feed=nil)
    if feed
      $db.execute("select * from items")
    else
      $db.execute("select * from items where feed_id = ?", feed)
    end
  end

  def mark_as_unread(item)
    $db.execute("update items set read = 0 where id = #{item['id']}")
  end

  def mark_as_read(item)
    $db.execute("update items set read = 1 where id = #{item['id']}")
  end
end

class Screen
  def initialize(stdscr)
    @stdscr = stdscr

    @min_y = 1
    @max_y = @stdscr.getmaxy

    @cur_y = 1
    @cur_x = 0
  end

  def display_menu
    @stdscr.clear()
    @stdscr.move(0, 0)
    @stdscr.addstr(" rutt #{@menu}\n")
  end

  def move_pointer(pos, move_to=false)
    @stdscr.move(@cur_y, 0)
    @stdscr.addstr(" ")

    if move_to == true
      @cur_y = pos
    else
      @cur_y += pos
    end

    @stdscr.move(@cur_y, 0)
    @stdscr.addstr(">")
  end
end


class FeedScreen < Screen
  def initialize(stdscr)
    @feeds = {}
    @menu = "q:Quit d:delete r:refresh all"

    reload_feeds

    super(stdscr)
  end

  def reload_feeds
    @_feeds = Feed::all
  end

  def display_feeds
    @cur_y = @min_y

    @_feeds[@limit[0]..@limit[1]].each do |feed|

#      next if feed.unread == 0

      @stdscr.move(@cur_y, 0)
      @stdscr.addstr("  #{feed['unread']}/#{feed['num_items']}\t\t#{feed['title']}\n")
      @feeds[@cur_y] = feed

      @cur_y += 1
    end

    @cur_y = @min_y
    @stdscr.refresh
  end

  def window(start_limit=nil, end_limit=nil)
    @stdscr.clear

    @limit = [start_limit, end_limit] if start_limit || end_limit

    reload_feeds
    display_menu
    display_feeds
    move_pointer(0)
  end

  def loop
    window(0, @stdscr.getmaxy - 2)

    while true do
      c = @stdscr.getch

      if c > 0 && c < 255
        case c.chr
        when /q/i
          break
        when /a/i
          # no-op
        when /r/i
          cur_y = @cur_y

          Feed::refresh

          window
          move_pointer(cur_y, move_to=true)
        when /d/i
          cur_y = @cur_y

          @stdscr.clear
          display_menu
          feed = @feeds[cur_y]
          @stdscr.move(2, 0)
          @stdscr.addstr("Are you sure you want to delete #{feed['title']}? ")
          d = @stdscr.getch
          if d.chr =~ /y/i
            Feed::delete(feed)
            window
            move_pointer(cur_y, move_to=true)
          end
        when /p/i
          window(@limit[0] - @stdscr.getmaxy - 2, @limit[0])
        when /n/i
          window(@limit[1], @limit[1] + @stdscr.getmaxy - 2)
        when / /
          cur_y = @cur_y
          item_screen = ItemScreen.new(@stdscr, @feeds[cur_y])
          item_screen.loop

          window
          move_pointer(cur_y, move_to=true)
        end
      else
        case c
        when Ncurses::KEY_UP
          move_pointer(-1)
        when Ncurses::KEY_DOWN
          move_pointer(1)
        end
      end
    end
  end
end

class ItemScreen < Screen
  def initialize(stdscr, feed)
    @feed = feed
    @menu = " i:quit r:refresh m:mark as read u:mark as unread a:mark all as read b:open in browser"

    super(stdscr)
  end

  def reload_items
    @items = Item::all(@feed)
  end

  def display_items
    @cur_y = @min_y

    @items[@limit[0]..@limit[1]].each do |item|
      @stdscr.addstr("  #{item['read'].to_i == 0 ? 'N' : ' '}\t#{item['published_at']}\t#{item['title']}\n")

      @items[@cur_y] = item
      @cur_y += 1
    end

    @cur_y = @min_y
    @stdscr.refresh
  end

  def window(start_limit=nil, end_limit=nil)
    @stdscr.clear

    @limit = [start_limit, end_limit] if start_limit || end_limit

    reload_items
    display_menu
    display_items
    move_pointer(0)
  end

  def loop
    window(0, @stdscr.getmaxy - 2)

    while true do
      c = @stdscr.getch

      if c > 0 && c < 255
        case c.chr
        when /[iq]/i
          break
        when /a/i
          Feed::mark_as_read(@feed)
          window
          move_pointer(@cur_y, move_to=true)
        when /p/i
          window(@limit[1] - @stdscr.getmaxy - 2, @limit[1])
        when /n/i
          window(@limit[1], @limit[1] + @stdscr.getmaxy - 2)
        when /b/i
          cur_y = @cur_y
          Item::mark_as_read(@items[cur_y])
          Launchy.open(@items[cur_y]['url'])
        when /m/i
          cur_y = @cur_y
          Item::mark_as_read(@items[cur_y])
          window
          move_pointer(cur_y + 1, move_to=true)
        when /u/i
          cur_y = @cur_y
          Item::mark_as_unread(@items[cur_y])
          window
          move_pointer(cur_y + 1, move_to=true)
        when /r/i
          Feed::refresh_for(@feed)
          window
        when / /
          cur_y = @cur_y

          content_screen = ContentScreen.new(@stdscr, @items[@cur_y])
          content_screen.loop

          window
          move_pointer(cur_y, move_to=true)
        end
      else
        case c
        when Ncurses::KEY_UP
          move_pointer(-1)
        when Ncurses::KEY_DOWN
          move_pointer(1)
        end
      end
    end
  end
end





class ContentScreen < Screen
  def initialize(stdscr, item)
    @item = item
    @menu = "i:back b:open in browser"

    super(stdscr)
  end

  def get_content
    @content = `elinks -dump -dump-charset ascii -force-html #{@item['url']}`
  end

  def move_pointer(pos)
    return if @cur_line + pos < 0

    @stdscr.addstr("#{@item['title']} (#{@item['url']})\n")
    @cur_line += pos

    lines = @content[@cur_line..@cur_line + @stdscr.getmaxy - 5]
    cur_y = 2

    while cur_y < @cur_line + @stdscr.getmaxy - 5 do
      @stdscr.addstr("#{lines[cur_y]}\n")
      cur_y += 1
    end

    @stdscr.refresh
  end

  def window(pointer_pos)
    @stdscr.clear()
    display_menu()
    move_pointer(pointer_pos)
  end

  def loop
    @cur_line = 0
    get_content()
    window(0)

    while true do
      c = @stdscr.getch()
      if c > 0 && c < 255
        case c.chr
        when /iq/i
          Item::mark_as_read(@item)
          return
        when /b/i
          Launchy.open(@item['url'])
        when / /i
          window(10)
        end
      else
        case c
        when Ncurses::KEY_UP
          window(-1)
        when Ncurses::KEY_DOWN
          window(1)
        end
      end
    end

  end
end



def start_screen
  stdscr = Ncurses.initscr()

  Ncurses.start_color();
  Ncurses.cbreak();
  Ncurses.noecho();
  Ncurses.keypad(stdscr, true);

  # Initialize few color pairs
  # Ncurses.init_pair(1, COLOR_RED, COLOR_BLACK);
  # Ncurses.init_pair(2, COLOR_BLACK, COLOR_WHITE);
  # Ncurses.init_pair(3, COLOR_BLACK, COLOR_BLUE);
  # stdscr.bkgd(Ncurses.COLOR_PAIR(2));

  stdscr.clear

  feed_screen = FeedScreen.new(stdscr)
  feed_screen.loop
ensure
  Ncurses.endwin()
end


def main
  Feed::make_table!
  Item::make_table!

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: rutt.rb [options]"

    opts.on('-a', '--add FEED', "Add a feed") do |url|
      Feed::add(url)
      exit
    end

    opts.on('-r', '--refresh', "Refresh feeds.") do
      Feed::refresh
      exit
    end

    opts.on('-o', '--import-opml FILE', "Import opml") do |file|
      urls = Opml::get_urls(file)

      urls.each do |url|
        puts "Adding #{url}"
        Feed::add(url, false)
      end

      exit
    end

#    opts.on('-l', '--list-feeds', action='store_true', help="List the feeds")

  end.parse!


  start_screen
end

main
